### ISSUE TITLE
Updated vector not found by get_vector_by_id

### ISSUE THREAD

**Issue by CrazyBeeline** [milvus-io/milvus#1900]
Hello, dear Milvus team
I need to update features often

Some features must be deleted (x), some new features need to be inserted (y), and some features need to be updated (z)

My approach is to delete (x + y + Z) and insert (y + Z)

When I did this for about a million features, I found that the number of features in Milvus did not match the number I needed

I compared the features in Milvus with the features I needed, and found that some features in Milvus were not deleted. Some of the features I need are not in Milvus

My operation was carried out in batches, so I checked each batch.
In each batch, The deleted feature Milvus could not be found. The inserted feature Milvus could also be found
I don't know what happened. I hope Milvus can help me

milvus : 0.7.1
deleteOPS : java-sdk
insertOPS : resltful api
collectionInfo :
![image](https://user-images.githubusercontent.com/25030234/79028956-d96fa300-7bc4-11ea-9862-cdbbcbfc61eb.png)
cpu or gpu : cpu

**Comment by yhmo** [comment]
Please provide more detail steps for your problem. How many times you call insert/delete? How many vectors were delete/insert in an operation? Did you call flush() before get_vector_by_id()?

**Comment by CrazyBeeline** [comment]
In a batch, I will delete 100 to 200 features and insert more than 4000 features

I will flush every time I delete and insert a feature

About 200 batches operated

**Comment by yhmo** [comment]
A: to be deleted vectors
B: to be updated vectors
C: new vectors

1. delete 100~200 vectors(A+B)
2. flush
3. insert 4000 vectors(B+C)
4. flush

Do step 1~4 for 200 times.
After each step3 operation, some deleted vectors still can be found by get_vector_by_id, right?
After each step4 operation, some inserted vectors could not be found by get_vector_by_id, right?

**Comment by CrazyBeeline** [comment]
I mean！！
A: to be deleted vectors
B: to be updated vectors
C: new vectors

delete 4100~4200 vectors(A+B+C)
flush
insert 4000 vectors(B+C)
flush
Do step 1~4 for 200 times.
After each step3 operation,  deleted vectors can not be found by get_vector_by_id
After each step4 operation,  inserted vectors could  be found by get_vector_by_id

After 200 times,some deleted vectors still can be found by get_vector_by_id
After 200 times,some inserted vectors could not be found by get_vector_by_id

**Comment by JinHai-CN** [comment]
@CrazyBeeline I also can't clearly understand your steps. I would like to write down my understanding:

loop once {
    delete_vector(100~200);
    flush();
    add_vector(4000);
    flush();
}
Use GetVectorByID to get vector, the result is correct.

loop 200 times {
    delete_vector(100~200);
    flush();
    add_vector(4000);
    flush();
}
Use GetVectorByID to get vector, the result is wrong. Some deleted vectors are found and inserted vectors are not found.

Besides the steps. Have you check the log? Is there any thing found?

**Comment by CrazyBeeline** [comment]
@JinHai-CN  
I'm sorry to reply you so late
In my business, I need to delete some useless vectors, update some expired vertors and insert some new vertors
------------------------------------------------------------------------
ONE BATCH
A: to be deleted vectors (A about 100-200 each batch)
B: to be updated vectors 
C: new vectors ( B+C about 4000 each batch)
I know which vectors to delete(A), but I don't know which vectors need to be updated (B)and which vectors are new(C)
Fortunately, I know the latest vectors(B+C)
So delete all vectors(A+B+C), insert the latest vectors(B+C)
batch {
delete_vector(4100~4200)(A+B+C);
flush();
add_vector(4000)(B+C);
flush();
}
Each batch uses GetVectorByID to get all the vectors（A+B+C）
The deleted vectors (A) cannot be obtained, and the inserted vectors(B+C) can  be obtained.
The result is correct
-------------------------------------------------------------------------
Repeat the above 200 times, I found that the number of Milvus vectors didn't match the reality
So I analyzed the vectors of Milvus
It is found that some vectors have not been deleted and some vectors that need to exist do not exist
There is no problem with the operation of each batch, but the final result is not correct
At the same time, I checked the log and found no problem information
--------------------------------------------------------------------------
I don't know if I can describe the problem clearly
If there's something else that hasn't been described clearly, maybe we can communicate by phone

**Comment by yhmo** [comment]
I try reproduce this issue with the following code, but could not reproduce it:

The following code firstly insert 4000 vectors into a new collecton, and then repeat to delete 200 vectors, update 1000 vecetors, insert 4000 new vectors for 200 times. And check existence of deleted/updated item in each time.
```
# 3rd imports
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.abstract import CollectionSchema, IndexParam
from milvus.client.types import MetricType

TABLE_NAME = "AAA"
TABLE_DIMENSION = 128
BATCH_COUNT = 4000

def gen_vec_list(nb, seed=1234):
    ran = np.random.RandomState(seed)
    xb = ran.rand(nb, TABLE_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def count_table(milvus=None):
    status, row_count = milvus.count_collection(collection_name=TABLE_NAME)
    print(TABLE_NAME, "row count =", row_count)
    return status

def flush_table(milvus=None):
    milvus.flush(collection_name_array=[TABLE_NAME])

def main():
    milvus = Milvus()
    status = milvus.connect(host='127.0.0.1', port='19530')
    print("Connect to milvus status: ", status)

    status = milvus.create_collection({
        'collection_name': TABLE_NAME,
        'dimension': TABLE_DIMENSION,
        'index_file_size': 128,
        'metric_type': MetricType.L2
    })
    print("Create table status:", status)

    for k in range(200):
        print("No.", k)
        if k == 0:
            # add 0~4000
            print('add new vectors')
            new_ids = [k * BATCH_COUNT + i for i in range(BATCH_COUNT)]
            vec_list = gen_vec_list(BATCH_COUNT, k*13)
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=vec_list, ids=new_ids)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
        else:
            # delete 0~`100, 2000~2100
            print('delete vectors')
            delete_ids_1 = [(k-1) * BATCH_COUNT + i for i in range(100)]
            milvus.delete_by_id(collection_name=TABLE_NAME, id_array=delete_ids_1)
            delete_ids_2 = [(k-1) * BATCH_COUNT + i + 2000 for i in range(100)]
            milvus.delete_by_id(collection_name=TABLE_NAME, id_array=delete_ids_2)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
            # update 1000~1500, 3000~3500
            print('delete update vectors')
            update_ids_1 = [(k-1) * BATCH_COUNT + i + 1000 for i in range(500)]
            milvus.delete_by_id(collection_name=TABLE_NAME, id_array=update_ids_1)
            update_ids_2 = [(k-1) * BATCH_COUNT + i + 3000 for i in range(500)]
            milvus.delete_by_id(collection_name=TABLE_NAME, id_array=update_ids_2)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
            # add update 1000~1500, 3000~3500
            print('add update vectors')
            vec_list = gen_vec_list(500, k*17)
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=vec_list, ids=update_ids_1)
            vec_list = gen_vec_list(500, k*11)
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=vec_list, ids=update_ids_2)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
            # add new 4000
            print('add new vectors')
            vec_list = gen_vec_list(BATCH_COUNT, k * 23)
            new_ids = [k * BATCH_COUNT + i for i in range(BATCH_COUNT)]
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=vec_list, ids=new_ids)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)

            # check deleted
            for d in delete_ids_1:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                # print(status, vec)
                if len(vec) > 0:
                    print("ERROR! deleted vector_1 should not be found")
            for d in delete_ids_2:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) > 0:
                    print("ERROR! deleted vector_2 should not be found")

            # check updated
            for d in update_ids_1:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! updated vector_1 should be found")
            for d in update_ids_2:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! updated vector_2 should be found")

            # check new
            for d in new_ids:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! new vector should be found")

    status = milvus.disconnect()


if __name__ == '__main__':
    main()
```

**Comment by CrazyBeeline** [comment]
dear  @yhmo 
Thank you very much for your reply！
Every batch result I tested is also correct
I want to know if the final number of vectors is 760200？


If the total is correct, my different operation from yours is:
I don't know which vectors need to be updated and which vectors are new,so i do it below:
           # delete 0~`100, 2000~2100 delete 1000~1500, 3000~3500 delete new 4000
            delete_ids_1 = [(k-1) * BATCH_COUNT + i for i in range(100)]
            delete_ids_2 = [(k-1) * BATCH_COUNT + i + 2000 for i in range(100)]
            update_ids_1 = [(k-1) * BATCH_COUNT + i + 1000 for i in range(500)]
            update_ids_2 = [(k-1) * BATCH_COUNT + i + 3000 for i in range(500)]
            new_ids = [k * BATCH_COUNT + i for i in range(BATCH_COUNT)]
            milvus.delete_by_id(collection_name=TABLE_NAME, id_array=(delete_ids_1+delete_ids_2+update_ids_1+update_ids_2+new_ids)
		    flush_table(milvus=milvus)
           
		   #insert 1000~1500, 3000~3500 insert new 4000
	    vec_list1 = gen_vec_list(500, k*17)
            vec_list3 = gen_vec_list(500, k*11)
            vec_list2 = gen_vec_list(BATCH_COUNT, k * 23)
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=(vec_list1+vec_list2+vec_list3, ids=(update_ids_1+update_ids_2+new_ids)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
			# check deleted
			# check insert

Vector latitude and index type are different
![image](https://user-images.githubusercontent.com/25030234/79090878-b63c2380-7d7d-11ea-8d2d-43057bb5c3fc.png)


Look forward to your reply！！！

**Comment by yhmo** [comment]
I modify the code to test again:
1. create a new collection, index_file_size = 1024, metric=IP
2. assign index to the collection, type=IVFSQ8, nlist=1000
3. insert 4000 vectors id from 0~3999
4. delete A(delete 200)+B(update 1000)+C(new 4000), flush, insert B+C, flush, do 200 times
5. check existence for deleted/updated/new vectors for each time

Run this script will cost about 1 hour, so far no error encountered.

```
# 3rd imports
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.abstract import CollectionSchema, IndexParam
from milvus.client.types import MetricType

TABLE_NAME = "AAA"
TABLE_DIMENSION = 128
INDEX_TYPE = IndexType.IVF_SQ8
INDEX_PARAM = {'nlist': 16384}

BATCH_COUNT = 4000

def gen_vec_list(nb, seed=1234):
    ran = np.random.RandomState(seed)
    xb = ran.rand(nb, TABLE_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def count_table(milvus=None):
    status, row_count = milvus.count_collection(collection_name=TABLE_NAME)
    print(TABLE_NAME, "row count =", row_count)
    return status

def flush_table(milvus=None):
    milvus.flush(collection_name_array=[TABLE_NAME])

def main():
    milvus = Milvus()
    status = milvus.connect(host='127.0.0.1', port='19530')
    print("Connect to milvus status: ", status)

    status = milvus.create_collection({
        'collection_name': TABLE_NAME,
        'dimension': TABLE_DIMENSION,
        'index_file_size': 128,
        'metric_type': MetricType.IP
    })
    print("Create table status:", status)
    status = milvus.create_index(TABLE_NAME, index_type=INDEX_TYPE, params=INDEX_PARAM)

    for k in range(200):
        print("No.", k)
        if k == 0:
            # add 0~4000
            print('add new vectors')
            new_ids = [k * BATCH_COUNT + i for i in range(BATCH_COUNT)]
            vec_list = gen_vec_list(BATCH_COUNT, k*13)
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=vec_list, ids=new_ids)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)
        else:
            # A: delete 0~`100, 2000~2100 id array
            delete_ids_1 = [(k-1) * BATCH_COUNT + i for i in range(100)]
            delete_ids_2 = [(k-1) * BATCH_COUNT + i + 2000 for i in range(100)]
            # B: update 1000~1500, 3000~3500 id array
            update_ids_1 = [(k-1) * BATCH_COUNT + i + 1000 for i in range(500)]
            update_ids_2 = [(k-1) * BATCH_COUNT + i + 3000 for i in range(500)]
            # C: new 4000 id array
            new_ids = [k * BATCH_COUNT + i for i in range(BATCH_COUNT)]
            # delete A+B+C
            print('delete vectors')
            all_ids = new_ids + delete_ids_1 + delete_ids_2 + update_ids_1 + update_ids_2
            status = milvus.delete_by_id(collection_name=TABLE_NAME, id_array=all_ids)
            if status.code != 0:
                print("ERROR! delete_by_id return", status.message)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)

            # add 1000~1500, 3000~3500, and new 4000
            print('add vectors')
            update_vec_1 = gen_vec_list(500, k * 17)
            update_vec_2 = gen_vec_list(500, k * 23)
            new_vec = gen_vec_list(BATCH_COUNT, k*11)
            all_vec = new_vec + update_vec_1 + update_vec_2
            all_ids = new_ids + update_ids_1 + update_ids_2
            status, ids = milvus.insert(collection_name=TABLE_NAME, records=all_vec, ids=all_ids)
            if status.code != 0:
                print("ERROR! insert return", status.message)
            flush_table(milvus=milvus)
            count_table(milvus=milvus)

            # check deleted
            for d in delete_ids_1:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                # print(status, vec)
                if len(vec) > 0:
                    print("ERROR! deleted vector_1 should not be found")
            for d in delete_ids_2:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) > 0:
                    print("ERROR! deleted vector_2 should not be found")

            # check updated
            for d in update_ids_1:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! updated vector_1 should be found")
            for d in update_ids_2:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! updated vector_2 should be found")

            # check new
            for d in new_ids:
                status, vec = milvus.get_vector_by_id(collection_name=TABLE_NAME, vector_id=d)
                if len(vec) == 0:
                    print("ERROR! new vector should be found")

    status = milvus.disconnect()


if __name__ == '__main__':
    main()
```

**Comment by yhmo** [comment]
I got an error at No.198 batch:

```
No. 197
delete vectors
AAA row count = 747600
add vectors
AAA row count = 752600
No. 198
delete vectors
AAA row count = 751400
add vectors
AAA row count = 756400
ERROR! updated vector_2 should be found
No. 199
delete vectors
AAA row count = 755200
add vectors
AAA row count = 760200
```

**Comment by yhmo** [comment]
Treat as a bug and start investigate.

**Comment by CrazyBeeline** [comment]
Thank God for repeating the question
In my test, I found that some vectors were not deleted, not only the updated vectors could not be found
Please kindly be informed

**Comment by yhmo** [comment]
@CrazyBeeline, thank you for point out this issue. It is an important issue, we will fix it in 0.8.0.

**Comment by yhmo** [comment]
Fixed by PR1923, in 0.8.0

**Comment by CrazyBeeline** [comment]
@yhmo 
Version 0.8.0 repeats this problem

**Comment by yhmo** [comment]
@CrazyBeeline 
Do you mean "deleted vectors still can be found"?

**Comment by CrazyBeeline** [comment]
@yhmo  yes

**Comment by CrazyBeeline** [comment]
Version 0.9.0 repeats this problem
some vertex deleted can find by get_vector_by_id 
@yhmo

**Comment by JinHai-CN** [comment]
@CrazyBeeline Which meta database are you using, MySQL or SQLite?

**Comment by CrazyBeeline** [comment]
mysql

**Comment by tinkerlin** [comment]
You can retest on the latest release 0.9.1.
we can't reproduce the same problem locally. Hopes the latest release may help you.
@CrazyBeeline

**Comment by CrazyBeeline** [comment]
I updated milvus to 0.9.1
Deleted vectors can still be retrieved
It's strange that the total number of vectors is correct
I tested ivfsq8 ivfsq8h with GPU

@tinkerlin

**Comment by CrazyBeeline** [comment]
Update last comment
After a lot of vectors testing with milvus 0.9.1 
Deleted vectors can be retrieved, inserted vectors cannot be searched

**Comment by yhmo** [comment]
@CrazyBeeline 

To ensure inserted vectors can be searched immediately, you need to call flush(collection_name) to flush the vectors into disk storage. Or they will be searchable after a few seconds, depend on the 'auto_flush_interval' in the server_config.yaml, this parameter default value is 1 second, means background thread automatically flush vectors into disk per second.
`auto_flush_interval: 1`

**Comment by yhmo** [comment]
For "Deleted vectors can be retrieved", could you provide some python script that can reproduce it?

**Comment by CrazyBeeline** [comment]
I will provide correlation vectors and algorithms next week
@yhmo

**Comment by CrazyBeeline** [comment]
链接：https://pan.baidu.com/s/1VN27HHBDtgMtk0SEx216hA 
提取码：x4r0 
@yhmo

**Comment by yhmo** [comment]
get it

**Comment by yhmo** [comment]
Steps to reproduce "Deleted vectors can be retrieved" in 0.10.0:
1. set "auto_flush_interval = 3600" in server_config.yaml
2. set "insert_buffer_size = 10MB" in server_config.yaml
3. start milvus server
4. create a collection with 512 dimension
5. insert 10000 vectors with the following code:
```
ids = [k for k in range(10000)]
vec_list = gen_vec_list(10000)
MILVUS.insert(collection_name=TABLE_NAME, records=vec_list, ids=ids)
```
6. delete some vectors
```
delete_ids = [k for k in range(100)]
MILVUS.delete_entity_by_id(collection_name=TABLE_NAME, id_array=delete_ids)
```
7. insert more vectors to hit the "insert_buffer_size" threshold, internal flush() is triggered
```
ids = [k+20000 for k in range(20000)]
vec_list = gen_vec_list(20000)
MILVUS.insert(collection_name=TABLE_NAME, records=vec_list, ids=ids)
```
8. invoke flush(), but it does nothing since the internal flush() has been executed
9. get vector which has been deleted
`status, data = MILVUS.get_entity_by_id(collection_name=TABLE_NAME, ids=[10])`
The vector with id '10' should be deleted, but it still can be retrieved

**Comment by yhmo** [comment]
The root cause is in milvus/core/src/db/insert/MemManagerImpl.cpp, line No.39
```
if (GetCurrentMem() > options_.insert_buffer_size_) {
        // TODO(zhiru): Don't apply delete here in order to avoid possible concurrency issues with Merge
        auto status = Flush(flushed_tables, false);
        fiu_do_on("MemManagerImpl::InsertVectors_flush_fail", status = Status(milvus::SERVER_UNEXPECTED_ERROR, ""));
        if (!status.ok()) {
            return status;
        }
    }
```
When the insert buffer hit "insert_buffer_size" threshold, the internal flush() method parameter "apply_delete=false"

**Comment by CrazyBeeline** [comment]
P1. I understand why I can retrieve deleted vectors
P2. Is the inconsistency of the total number of vectors caused by this problem?
P3. How can I avoid the problem in P1 and P2

**Comment by yhmo** [comment]
P2. I think yes, but not completely confirm.
P3. Maybe this way can avoid the problem : enlarge the "insert_buffer_size" to 2~4, and manually call flush() before the accumulated insert data hit the "insert_buffer_size"

**Comment by CrazyBeeline** [comment]
dear yhmo！
I don't think this solution will solve the problem
my environment insert_ buffer_ size:1G
Every time I insert  vectors, it will not exceed 100mb
Flush every time after insert  vectors

As you said, more in-depth research is needed
Looking forward to your next reply

**Comment by ThreadDao** [comment]
This issue should be resolved in 0.11.0, need check.

**Comment by ThreadDao** [comment]
I did not find any problems with version 0.11.1 by the yhmo's scripts. And I will close this issue. @CrazyBeeline you can reopen it if you reproduce the problem on 0.11.1. Thank you
